zonemaster/zonemaster-backend
# Properly define properties to use for database fingerprint (deterministic hash)

## Purpose

When we store a test in the database, we also store a fingerprint based on some criteria, that can be used to avoid running the same time in a short period of time.
This PR harmonizes the way we compute the fingerprint so that we can reuse tests that are similar, even if some properties are left out by the client.
This also normalize the value stored in the `params` database field.

## Context

As explained in #775, depending on the client used, some properties are used in the computation of the fingerprint or not. This PR solves this by always using the same properties (falling back to a default value if the property is not set).
Fixes #775

## Changes

* a first commit is only about removing trailing spaces and changing tabs into spaces (839f435)
* update variable name "deterministic hash" into "fingerprint"
* create new methods to normalize the parameters and to compute the fingerprint
* always include the same properties to compute the fingerprint, only the value will differ
* use default value if the property is not already set (`profile: default` and look into the profile to get the value for `ipv4 ` and `ipv6`)
* normalize the `params` object to a minimal set enlarged with extra values given by the client (such as `client_id` and `client_version`)

## Breaking change

Removal of the `config` property from the config file.
(TBH, this breaking change has not been introduced with this PR, however this PR removes the documentation associated with this property)

## How to test this PR

Unit tests are added with this PR. It is also possible to test using the scenario explained in #775 or by using the `get_test_params` API call.

Testing with the following test case allows to make sure the database is properly populated and that the behavior has not been altered (you may need to increase the `age_reuse_previous_test` value in the config while running the following tests).

1. Starting twice the same test should give the same test-id (the diff output should be empty):
    ```
    TESTID1=$(zmb start_domain_test --domain example.com | jq -r '.result')
    TESTID2=$(zmb start_domain_test --domain example.com | jq -r '.result')
    diff <(echo $TESTID1) <(echo $TESTID2)
    ```

2. Adding garbage option should give the same test-id:
    ```
    TESTID3=$(zmb start_domain_test --domain example.com --client-id "zmb" | jq -r '.result')
    diff <(echo $TESTID1) <(echo $TESTID3)
    ```

3. Adding nameservers without ip should give another test-id (the diff should give an error):
    ```
    TESTID4=$(zmb start_domain_test --domain example.com --nameserver a.example.com | jq -r '.result')
    diff <(echo $TESTID1) <(echo $TESTID4)
    ```

4. Adding ip to the nameserver should give another test-id (both diff should fail):
    ```
    TESTID5=$(zmb start_domain_test --domain example.com --nameserver a.example.com:192.0.2.1 | jq -r '.result')
    diff <(echo $TESTID1) <(echo $TESTID5)
    diff <(echo $TESTID4) <(echo $TESTID5)
    ```

5. Wait `age_reuse_previous_test` and run:
    ```
    TESTID6=$(zmb start_domain_test --domain example.com --client-id "my personal client" | jq -r '.result')
    TESTID7=$(zmb start_domain_test --domain example.com --client-id "another client" | jq -r '.result')
    diff <(echo $TESTID6) <(echo $TESTID7)
    ```

6. Finally output each test parameters and check that the values are correct
    ```
    zmb get_test_params --test-id $TESTID1
    zmb get_test_params --test-id $TESTID2
    zmb get_test_params --test-id $TESTID3  # should not contain any "client_id" property
    zmb get_test_params --test-id $TESTID4
    zmb get_test_params --test-id $TESTID5
    zmb get_test_params --test-id $TESTID6  # should have a "client_id" property set to "my personal client"
    zmb get_test_params --test-id $TESTID7  # should have a "client_id" property which is NOT "another client" but "my personal client"
    ```
